README.md
<h1 align="center">
  <img src="./assets/logo.png" alt="secure-rm" width="25%">
  <br>
  <strong>Secure-rm</strong>
  <br>
</h1>
<p align="center">Data erasure solution for files and drives</p>

<p align="center">
  <a href="https://www.npmjs.com/package/secure-rm"><img src="https://img.shields.io/npm/v/secure-rm.svg?style=flat-square" alt="Version"></a>
  <a href="https://www.npmjs.com/package/secure-rm"><img src="https://img.shields.io/npm/dw/secure-rm.svg?style=flat-square" alt="Downloads/week"></a>
  <a href="https://github.com/oganexon/secure-rm/blob/master/LICENSE"><img src="https://img.shields.io/npm/l/secure-rm.svg?style=flat-square" alt="License: MIT"></a>
</p>
<p align="center">
  <a href="https://github.com/secure-rm/core/actions?query=branch%3Amaster"><img alt="GitHub Workflow Status (branch)" src="https://img.shields.io/github/workflow/status/secure-rm/core/Node%20CI/master?style=flat-square&label=master%20build" alt="Build Status: master"></a>
  <a href="https://github.com/secure-rm/core/actions?query=branch%3Adevelop"><img alt="GitHub Workflow Status (branch)" src="https://img.shields.io/github/workflow/status/secure-rm/core/Node%20CI/develop?style=flat-square&label=dev%20build" alt="Build Status: develop"></a>
  <a href="https://coveralls.io/github/secure-rm/core"><img src="https://img.shields.io/coveralls/github/secure-rm/core?style=flat-square" alt="Code coverage"></a>
</p>

## ❓ Why

Secure-rm is composed of two parts: a Node.js module with a straightforward API and a [command line interface](https://www.npmjs.com/package/secure-rm-cli) optimized to delete files on the fly.

When you delete a file using the `rm` UNIX command or `fs.unlink` in node, it only removes direct pointers to the data disk sectors and make the data recovery possible with common software tools.

Permanent data erasure goes beyond basic file deletion commands, which:

1. Allow for selection of a specific standard, based on unique needs,
2. Verify the overwriting method has been successful and removed data across the entire device.

## 🔩 How It Works

The basic principle is to write files before deletion in order to make recovery harder. With secure-rm, you get to choose the standard that follow your needs. Each one is composed of instructions about how many passes it should perform.

It goes from a simple pass of zeros to a 35 passes algorithm. Secure-rm comes with its own algorithm to ensure your data is safe:

* A pass of cryptographically strong pseudo-random data,
* The file is then renamed,
* And finally truncated to hide the file size.

## ✨ Features

* [Choose your standard](https://docs.secure-rm.com/core/standards)
* [Create your own standard](https://docs.secure-rm.com/core/custom-standard)
  * [on files](https://docs.secure-rm.com/core/custom-standard/unlink-methods)
  * [on directories](https://docs.secure-rm.com/core/custom-standard/rmdir-methods)
* [Use events to follow the progression with huge files](https://docs.secure-rm.com/core/events.md)

### 📚 [Documentation](https://docs.secure-rm.com/core/getting-started)

## 📦 Installation

[Node](https://nodejs.org/) and [npm](https://www.npmjs.com/) required.

```shell
$ yarn add secure-rm
# or
$ npm install secure-rm
```

## 🚀 Quick Start

If you want your application to delete specific files with a pass of cryptographically strong pseudo-random data, use one of these code snippets:

```javascript
const srm = require('secure-rm')

srm.remove('./folder/file.js').result
  .then(() => console.log('Files successfully deleted !'))
  .catch(console.error)

// OR

srm.remove('./folder/file.js', (err) => {
  if (err) throw err
  console.log('Files successfully deleted !')
})
```

## 📋 Examples

```javascript
const options = {
  standard: srm.standards.gutmann,
  maxBusyTries: 5
}

srm.remove('./data/file.js', options).result
  .then(({count, index}) => console.log(`Successfully deleted ${count} files: ${index}`))
  .catch(console.error)

const events = srm.remove('./trash/dir/', { standard: srm.standards.schneier }, (err, {count, index}) => {
  if (err) throw err
  console.log(`Successfully deleted ${count} files: ${index}`))
})

event.on('removed', (fileName) => console.log(fileName))
```

## 📜 Changelog / History

See the [changelog](/CHANGELOG.md) or [releases](https://github.com/oganexon/secure-rm/releases).

## 🏗️ Contributing

<p align="center">
  <a href="https://libraries.io/npm/secure-rm"><img src="https://img.shields.io/librariesio/release/npm/secure-rm?style=flat-square&logo=npm" alt="Dependencies"></a>
  <img src="https://img.shields.io/github/contributors/secure-rm/core?style=flat-square" alt="Contributors">
  <img src="https://img.shields.io/github/last-commit/secure-rm/core/develop?style=flat-square" alt="Last commit">
  <img src="https://img.shields.io/npm/collaborators/secure-rm?style=flat-square" alt="npm collaborators">
</p>
<p align="center">
  <a href="https://jestjs.io"><img src="https://img.shields.io/badge/-jest-99424f?style=flat-square&logo=jest" alt="Tested with Jest"></a>
  <a href="https://nodejs.org"><img src="https://img.shields.io/badge/-node-gray?style=flat-square&logo=node.js" alt="Node version"></a>
  <a href="https://www.typescriptlang.org/"><img src="https://img.shields.io/badge/-typescript-blue?style=flat-square&logo=typescript" alt="language"></a>
</p>

See [contributing guidelines](/CONTRIBUTING.md)

## Todo

[] Fix wipeDrive (removed from 5.0)

### Licensing

Icon library by [Icons8](https://icons8.com/).

This project is under [MIT License](/LICENSE).


CHANGELOG.md
# [5.0.0](https://github.com/secure-rm/core/compare/v4.0.0-2...v5.0.0) (2020-05-06)



# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!--Types of changes
### Added
### Changed
### Deprecated
### Removed
### Fixed
### Security -->

<!-- ## [Unreleased] -->

<!-- - `` function to clear free space on drive.
- `` function to properly format drive. -->

## [5.0.0] - 2020-06-05

### Added

- Support for 64bit files and files of any size.

### Changed

- `srm` function becomes `srm.remove`
- Global syntax.
- `secure` standard changes timestamps.
- Internal changes:
  - Functions are async.
  - Standards syntax.
- `preview` standard becomes `mark`.

### Fixed

- Queues up open and readdir calls, and retries them once something closes if there is an EMFILE error from too many file descriptors.


## [4.2.0] - 2019-11-02

### Changed

- `secure` standard now also rename folders.

### Fixed

- Unhandled errors.

## [4.1.0] - 2019-10-29

### Changed

- Events are now:
  - 'debug'
  - 'info'
  - 'notice'
  - 'warn'
  - 'error'

### Fixed

- Npm auto publisher bug.

## [4.0.1] - 2019-10-27

### Added

- Npm auto publisher.

## [4.0.0] - 2019-10-27

### Added
- Custom directory removal.
- More tests.
- Wiki.
- Preview standard.

### Changed
- Repository location.
- New custom standard syntax.

### Security
- Libraries update.

## [3.0.3] - 2019-08-30

### Fixed
- Default standard when none is provided. (Tests are now a priority)

## [3.0.2] - 2019-08-30

### Fixed
- Default standard when none is provided.

## [3.0.1] - 2019-08-30

### Security
- Bumped eslint-utils from 1.4.0 to 1.4.2 

## [3.0.0] - 2019-08-29

### Added
- Dependabot.

### Removed
- CLI support. See [secure-rm-cli](https://www.npmjs.com/package/secure-rm-cli)
- Unused packages.

## [2.0.1] - 2019-08-27

### Added
- Code coverage.

### Deprecated
- The CLI is deprecated, it moved over `secure-rm-cli`.

## [2.0.0] - 2019-08-26

### Added
- End message when the process is finished.
- Types definitions.
- Mute argument in cli.
- New badges in readme.
- Travis Continuous Integration.

### Changed
- All JavaScript files have been converted to TypeScript files.
- 'Method' is now 'standard' in the cli and in the lib.
- Custom standard has changed. See the doc.

### Soon deprecated
- CLI will be ported in its own package.

## [1.1.1] - 2019-08-20

### Added
- Comments on every file.
- Error recap at the end of the process.

## [1.1.0] - 2019-08-19

### Fixed
- Command line tool bugs with paths.

## [1.0.1] - 2019-08-16

Release of 1.0.0 (stable API)
(messed up 1.0.0 on npm so it's 1.0.1)

## [1.0.0-beta] - 2019-08-16

### Added
- More tests.

### Changed
- Documentation:
  - Windows advice.
  - Beautified the whole.

## [0.14.0] - 2019-08-15

### Added
- Changelog.
- Promises support.
- `options` argument in `srm` function:
  - `options` [\<Object\>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) (optional) :
    - `method` [\<String\>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type) : ID of the method (default: 'secure');
    - `customMethod` [\<Function\>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function) : your own method to remove a file (if specified, priority over `method`);
    - `maxBusyTries` [\<Number\>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type) : number of retries if an error occur;
    - `disableGlob` [\<Boolean\>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type) : allow or not file globbing (default: true).
- CLI args accordingly.

### Changed
- Methods IDs are now textual rather than numerical.
- Args for `srm(path, methodID)` are now `srm(path, options)`.
- Updated documentation.


lib/write.js
const fs = require('fs')
const path = require('path')
const events = require('events')
const crypto = require('crypto')
const { kMaxLength } = require('buffer')

const eventEmitter = new events.EventEmitter()

// Function to offset an array
const offset = (arr, offset) => [...arr.slice(offset), ...arr.slice(0, offset)]

// Object listing writing methods
const write = {
  // Needed to provide file size
  init: file => {
    return new Promise((resolve, reject) => {
      fs.stat(file, (err, stats) => {
        if (err) reject(err)
        else if (stats.size <= kMaxLength) {
          eventEmitter.emit('start', file)
          resolve({ fileSize: stats.size, file })
        } else {
          reject(Error('64bit files are not yet supported.'))
        }
      })
    })
  },
  random: (file, fileSize, passes) => {
    return new Promise((resolve, reject) => {
      const buffer = Buffer.alloc(fileSize)
      crypto.randomFill(buffer, (err, randomBuffer) => {
        if (err) reject(err)
        eventEmitter.emit('info', file, 'Writing random data ')
        fs.writeFile(file, randomBuffer, (err) => {
          if (err) reject(err)
          else if (passes > 1) {
            write.random(file, fileSize, passes - 1)
              .then(() => resolve({ fileSize, file }))
              .catch((err) => reject(err))
          } else resolve({ fileSize, file })
        })
      })
    })
  },
  zeroes: (file, fileSize, passes) => {
    return new Promise((resolve, reject) => {
      eventEmitter.emit('info', file, 'Writing zeroes ')
      fs.writeFile(file, Buffer.alloc(fileSize, 0), (err) => {
        if (err) reject(err)
        else if (passes > 1) {
          write.zeroes(file, fileSize, passes - 1)
            .then(() => resolve({ fileSize, file }))
            .catch((err) => reject(err))
        } else resolve({ fileSize, file })
      })
    })
  },
  ones: (file, fileSize, passes) => {
    return new Promise((resolve, reject) => {
      eventEmitter.emit('info', file, 'Writing ones ')
      fs.writeFile(file, Buffer.alloc(fileSize, 0b11111111), (err) => {
        if (err) reject(err)
        else if (passes > 1) {
          write.ones(file, fileSize, passes - 1)
            .then(() => resolve({ fileSize, file }))
            .catch((err) => reject(err))
        } else resolve({ fileSize, file })
      })
    })
  },
  // Write one byte on the whole file
  byte: (file, data, fileSize, passes) => {
    return new Promise((resolve, reject) => {
      eventEmitter.emit('info', file, `Writing 0x${data.toString(16)} `)
      fs.writeFile(file, Buffer.alloc(fileSize, data), (err) => {
        if (err) reject(err)
        else if (passes > 1) {
          write.byte(file, data, fileSize, passes - 1)
            .then(() => resolve({ fileSize, file }))
            .catch((err) => reject(err))
        } else resolve({ fileSize, file })
      })
    })
  },
  // Write an array of bytes on the whole file
  bytes: (file, dataArray, fileSize, passes) => {
    return new Promise((resolve, reject) => {
      eventEmitter.emit('info', file, `Writing ${dataArray.map(x => '0x' + x.toString(16))} `)
      const dataConverted = Uint8Array.from(dataArray)
      fs.writeFile(file, Buffer.alloc(fileSize, dataConverted), (err) => {
        if (err) reject(err)
        else if (passes > 1) {
          write.bytes(file, dataArray, fileSize, passes - 1)
            .then(() => resolve({ fileSize, file }))
            .catch((err) => reject(err))
        } else resolve({ fileSize, file })
      })
    })
  },
  // Write an array of bytes on the whole file
  // And then cycle through the array
  cycleBytes: (file, dataArray, fileSize, passes = 1) => {
    return new Promise((resolve, reject) => {
      eventEmitter.emit('info', file, `Writing ${dataArray.map(x => '0x' + x.toString(16))} `)
      const dataConverted = Uint8Array.from(dataArray)
      fs.writeFile(file, Buffer.alloc(fileSize, dataConverted), (err) => {
        if (err) reject(err)
        else if (passes < dataArray.length) {
          write.cycleBytes(file, offset(dataArray, 1), fileSize, passes + 1)
            .then(() => resolve({ fileSize, file }))
            .catch((err) => reject(err))
        } else resolve({ fileSize, file })
      })
    })
  },
  // Write start byte on the whole file
  // And then + increment until end byte
  incrementByte: (file, { start, end, increment }, fileSize) => {
    return new Promise((resolve, reject) => {
      eventEmitter.emit('info', file, `Writing 0x${start.toString(16)} `)
      fs.writeFile(file, Buffer.alloc(fileSize, start), (err) => {
        if (err) reject(err)
        else if (start + increment <= end) {
          write.incrementByte(file, { start: start + increment, end: end, increment }, fileSize)
            .then(() => resolve({ fileSize, file }))
            .catch((err) => reject(err))
        } else resolve({ fileSize, file })
      })
    })
  },
  randomByte: (file, fileSize, passes) => {
    return new Promise((resolve, reject) => {
      eventEmitter.emit('info', file, 'Writing random byte ')
      fs.writeFile(file, Buffer.alloc(fileSize, crypto.randomBytes(1)[0]), (err) => {
        if (err) reject(err)
        else if (passes > 1) {
          write.randomByte(file, fileSize, passes - 1)
            .then(() => resolve({ fileSize, file }))
            .catch((err) => reject(err))
        } else resolve({ fileSize, file })
      })
    })
  },
  // Write the binary complement
  complementary: (file, fileSize) => {
    return new Promise((resolve, reject) => {
      eventEmitter.emit('info', file, 'Reading file ')
      fs.readFile(file, (err, data) => {
        if (err) reject(err)
        else {
          eventEmitter.emit('info', file, 'Getting binary complement ')
          for (let i = 0, l = fileSize; i < l; i++) {
            data[i] = ~data[i]
          }
          eventEmitter.emit('info', file, 'Writing binary complement ')
          fs.writeFile(file, data, (err) => {
            if (err) reject(err)
            else resolve({ fileSize, file })
          })
        }
      })
    })
  },
  // Rename to random string
  rename: (file, fileSize) => {
    return new Promise((resolve, reject) => {
      const newName = crypto.randomBytes(9).toString('base64').replace(/\//g, '0').replace(/\+/g, 'a')
      const newPath = path.join(path.dirname(file), newName)
      eventEmitter.emit('info', file, `Renaming to ${newName} `)
      fs.rename(file, newPath, (err) => {
        if (err) reject(err)
        else resolve({ fileSize, file: newPath })
      })
    })
  },
  // Truncate to between 25% and 75% of the file size
  truncate: (file, fileSize) => {
    return new Promise((resolve, reject) => {
      eventEmitter.emit('info', file, 'Truncating ')
      const newSize = Math.floor((Math.random() * 0.5 + 0.25) * fileSize)
      fs.truncate(file, newSize, (err) => {
        if (err) reject(err)
        else resolve({ fileSize: newSize, file })
      })
    })
  },
  // End function: unlink the file pointer
  unlink: file => {
    return new Promise((resolve, reject) => {
      eventEmitter.emit('unlink', file)
      fs.unlink(file, (err) => {
        if (err) reject(err)
        else {
          eventEmitter.emit('done', file)
          resolve()
        }
      })
    })
  }
}

module.exports = { write, eventEmitter }
